api/src/main/java/org/eclipse/microprofile/config/ConfigValue.java

```diff
@@ -0,0 +1,135 @@
+ * Copyright (c) 2009-2017 Mark Struberg and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *
+ *******************************************************************************/
+package org.eclipse.microprofile.config;
+
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Accessor to a configured value.
+ * It follows a builder pattern.
+ *
+ * Accessing the configured value is finally done via {@link #get()} or {@link #getOptional()}
+ *
+ * @author <a href="mailto:struberg@apache.org">Mark Struberg</a>
+ * @author <a href="mailto:gpetracek@apache.org">Gerhard Petracek</a>
+ * @author <a href="mailto:rsmeral@jboss.org">Ron Smeral</a>
+ */
+public interface ConfigValue<T> {
+
+    /**
+     * Sets the type of the configuration entry to the given class and returns this builder.
+     * The default type of a ConfigValue is {@code String}.
+     *
+     * This will use the registered {@link org.eclipse.microprofile.config.spi.Converter} for the given type.
+     **
+     * @param clazz The target type
+     * @param <N> The target type
+     * @return This builder as a typed ConfigValue
+     */
+    <N> ConfigValue<N> as(Class<N> clazz);
+
+    /**
+     * Specify that a resolved value will get cached for a certain amount of time.
+     * After the time expires the next {@link #get()} will again resolve the value
+     * from the underlying {@link Config}.
+     *
+     * @param value the amount of the TimeUnit to wait
+     * @param timeUnit the TimeUnit for the value
+     *
+     * @return This builder
+     */
+    ConfigValue<T> cacheFor(long value, TimeUnit timeUnit);
+
+    /**
+     * Whether to evaluate variables in configured values.
+     * A variable starts with '${' and ends with '}', e.g.
+     * <pre>
+     * mycompany.some.url=${myserver.host}/some/path
+     * myserver.host=http://localhost:8081
+     * </pre>
+     * If 'evaluateVariables' is enabled, the result for the above key
+     * {@code "mycompany.some.url"} would be:
+     * {@code "http://localhost:8081/some/path"}
+     * @param evaluateVariables whether to evaluate variables in values or not
+     * @return This builder
+     */
+    ConfigValue<T> evaluateVariables(boolean evaluateVariables);
+
+
+    /**
+     * The specified lambda will get called on any config value changes.
+     *
+     * @return This builder
+     */
+    ConfigValue<T> onChange(ConfigChanged valueChange);
```

Comment:
@OndrejM it will be a synchronous callback within the same thread. Of course we should better specify how it works and how it behaves.

@Emily-Jiang A use case would be 

```Java
ConfigValue cfgPort = config.access("myprj.archive.port")
    .as(Integer.class)
    .cacheFor(2, TimeUnit.MINUTES)
    .onChange(::logConfigChange);
```

and logConfigChange will then log out that a value got changed.
Of course there might be other tasks which get triggered when the configured value flips.